src/clj/puppetlabs/services/jruby/jruby_puppet_service.clj

```diff
@@ -42,6 +42,13 @@
             (assoc :pool-context pool-context)
             (assoc :borrow-timeout (:borrow-timeout config))
             (assoc :event-callbacks (atom []))))))
+  (stop
+   [this context]
+   (let [{:keys [pool-context]} (tk-services/service-context this)
+         flush-complete? (promise)]
+     (jruby-agents/send-flush-pool-for-shutdown! pool-context flush-complete?)
+     @flush-complete?)
```

Comment:
Yeah, not sure.  Could take a super-long time depending upon what's going on within the server.  I'd probably be inclined to not try to put a timeout on this.  Otherwise, the server might be in an unknown state if we continue on with the rest of the shutdown process.